### PR TITLE
Make open method resolver pay for play

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Resolver is called into from delegate thunks, so we can't get rid of it, but to construct a resolver one needs to do some reflection first. Make the resolution logic statically depend on resolver being constructed.

cc @dotnet/ilc-contrib

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs [hunk 1]
```diff
@@ -61,6 +65,7 @@ public OpenMethodResolver(RuntimeTypeHandle declaringType, IntPtr codePointer, G
             _declaringType = declaringType.ToEETypePtr();
             _handle = handle;
             _readerGCHandle = readerGCHandle;
+            s_lazyGvmLookupForSlot = &TypeLoaderExports.GVMLookupForSlot;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:68]: Do we really need to set it here? The type loader call is under `if (_resolveType == GVMResolve)`, but we are setting the resolver to `OpenNonVirtualResolve` in this method. I would expect this pointer to be only set when we set `_resolveType = GVMResolve`.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs [hunk 2]
```diff
@@ -42,6 +44,7 @@ public OpenMethodResolver(RuntimeTypeHandle declaringTypeOfSlot, int slot, GCHan
             _handle = handle;
             _readerGCHandle = readerGCHandle;
             _nonVirtualOpenInvokeCodePointer = IntPtr.Zero;
+            s_lazyGvmLookupForSlot = &TypeLoaderExports.GVMLookupForSlot;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:47]: Same here.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:68]: We don't - my concern was that if someone adds a new ctor, how do we make it less likely they'll forget to initialize this static? Person adding it _might_ look at one of the other ctors. But I don't have a strong opinion on this.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs [hunk 3]
```diff
@@ -20,8 +20,10 @@ namespace Internal.Runtime.CompilerServices
     // 3) Use the ResolveMethod function to do the virtual lookup. This function takes advantage of
     //    a lockless cache so the resolution is very fast for repeated lookups.
     [ReflectionBlocked]
-    public struct OpenMethodResolver : IEquatable<OpenMethodResolver>
+    public unsafe struct OpenMethodResolver : IEquatable<OpenMethodResolver>
     {
+        private static delegate*<object, RuntimeMethodHandle, nint> s_lazyGvmLookupForSlot;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:25]:
```suggestion
        // Lazy initialized to point to the type loader method when the first `GVMResolve` resolver is created
        private static delegate*<object, RuntimeMethodHandle, nint> s_lazyGvmLookupForSlot;
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs (hunk 1 above)
Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:68]: This risk can be mitigated by a comment (suggested one).

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs [hunk 4]
```diff
@@ -52,6 +55,7 @@ public unsafe OpenMethodResolver(RuntimeTypeHandle declaringTypeOfSlot, RuntimeM
             _handle = handle;
             _readerGCHandle = readerGCHandle;
             _nonVirtualOpenInvokeCodePointer = IntPtr.Zero;
+            s_lazyGvmLookupForSlot = &TypeLoaderExports.GVMLookupForSlot;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/OpenMethodResolver.cs:58]:
```suggestion
            if (s_lazyGvmLookupForSlot == null)
                s_lazyGvmLookupForSlot = &TypeLoaderExports.GVMLookupForSlot;
```
Avoid cache trashing?

**Pull request merged**